internal/premis/premis.go

```diff
@@ -0,0 +1,465 @@
+</premis:premis>
+
+`
+
+type PREMISObjectEventIdentifier struct {
+	IdType  string
+	IdValue string
+}
+
+type PREMISObject struct {
+	IdType           string
+	IdValue          string
+	OriginalName     string
+	EventIdentifiers []PREMISObjectEventIdentifier
+}
+
+type PREMISEventSummary struct {
+	Type    string
+	Detail  string
+	Outcome string
+}
+
+type PREMISEvent struct {
+	IdType       string
+	IdValue      string
+	Summary      PREMISEventSummary
+	DateTime     string
+	AgentIdType  string
+	AgentIdValue string
+}
+
+type PREMISAgent struct {
+	IdType  string
+	IdValue string
+	Name    string
+	Type    string
+}
+
+func PREMISAgentDefault() PREMISAgent {
+	return PREMISAgent{
+		Type:    "software",
+		Name:    "Enduro",
+		IdType:  "url",
+		IdValue: "https://github.com/artefactual-sdps/preprocessing-base",
+	}
+}
+
+func ReadBytesIfFileExists(filePath string) ([]byte, error) {
+	// Get file info.
+	_, err := os.Stat(filePath)
+
+	// Load bytes from file or initialize them.
+	var bytes []byte
+	if errors.Is(err, os.ErrNotExist) {
+		bytes = []byte("")
+	} else {
+		bytes, err = os.ReadFile(filepath.Clean(filePath))
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return bytes, nil
+}
+
+func writePREMIS(doc *etree.Document, writer io.Writer) error {
+	doc.Indent(2)
+
+	xml, err := doc.WriteToString()
+	if err != nil {
+		return err
+	}
+
+	_, err = writer.Write([]byte(xml))
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func ParsePREMISXmlAndGetRoot(xml string) (*etree.Document, *etree.Element, error) {
+	// Populate if empty.
+	if xml == "" {
+		xml = EmptyPremis
+	}
+
+	// Parse PREMIS document and get root PREMIS element.
+	doc := etree.NewDocument()
+
+	if err := doc.ReadFromString(xml); err != nil {
+		return nil, nil, err
+	}
+
+	PREMISEl := doc.FindElement("/premis:premis")
+	if PREMISEl == nil {
+		return nil, nil, errors.New("no root premis element found in document")
+	}
+
+	return doc, PREMISEl, nil
+}
+
+func BlankElementText(doc *etree.Document, xpath string) error {
+	el := doc.FindElement(xpath)
+	if el == nil {
+		return errors.New("element not found")
+	}
+
+	el.SetText("")
+
+	return nil
+}
+
+func AssembleEventDetailAndOutcome(failures []string) (string, string) {
+	detail := ""
+	outcome := "valid"
+
+	if failures != nil {
+		// Add failure descriptions to PREMIS event detail.
+		for _, description := range failures {
+			detail = detail + description + "\n"
+		}
+
+		outcome = "invalid"
+	}
+
+	return detail, outcome
+}
+
+func AppendPREMISObjectXML(xml string, writer io.Writer, object PREMISObject) error {
+	doc, PREMISEl, err := ParsePREMISXmlAndGetRoot(xml)
+	if err != nil {
+		return err
+	}
+
+	addPREMISObjectElementIfNeeded(PREMISEl, object)
+
+	err = writePREMIS(doc, writer)
```

Comment:
Again, it would be idiomatic to return the writer here, and then write the actual file in the activity.

Reply:
I rejiggered it so the Append* functions take an `etree.Document` as input and don't return anything (just alter the XML document) and the activities handle the writing to a file.